feat: 검색 기능 구현

devjaeyoon opened the pull request:
### ✨ 이슈 번호

- #76 

### 📌 설명

검색어를 입력하면 검색어에 해당하는 채널을 보여주는 기능을 구현한 Pull Request입니다.

### 📃 작업 사항

- [x] 검색 기능 구현
  - [x] 대소문자 구분없이 일관된 검색 및 결과를 보여줄 수 있도록 `toUpperCase` 사용
  - [x] 초기 어떠한 값도 입력하지 않았을 경우 `원하는 키워드(예시: SBS, 라디오, FM)를 입력해주세요.` 문구 출력
  - [x] 입력한 검색어에 대한 해당하는 결과가 없을 경우 `검색 결과가 없습니다.` 문구 출력  

### 💭 리뷰 사항 반영

- [x] 소문자 영어가 포함된 라디오 채널명에 대응하기 위해 `name`에 `.toUpperCase()` 적용
- [x] `const trimmedKeyword = keyword.trim().toUpperCase();`를 선언하여 중복 제거

### ✅ Pull Request 체크 사항

- [x] 가장 최신 브랜치를 pull 하였습니다.
- [x] base 브랜치명을 확인하였습니다.
- [x] 코드 컨벤션을 모두 확인하였습니다.
- [x] 브랜치명을 확인하였습니다.

devjaeyoon committed:
feat: 검색 키워드 전역 상태 관리용 useSearchStore 생성

devjaeyoon committed:
refactor: 검색 입력값을 useSearchStore 전역 상태로 관리하도록 변경

devjaeyoon committed:
feat: 라디오 채널 검색 결과 전용 SearchResultListItem 컴포넌트 생성

- 검색 페이지에 특화된 디자인 및 기능을 제공
- 채널 썸네일, 이름, 즐겨찾기 버튼 포함
- 리스트 클릭 시 플레이어로 이동, 별 클릭 시 즐겨찾기 토글

devjaeyoon committed:
feat: Search 컴포넌트에 라디오 채널 검색 기능 구현

- 키워드에 따라 라디오 채널 목록 필터링
- 키워드 미입력 시 안내 메시지 표시
- 검색 결과 없을 시 검색 결과 없음 메시지 노출
- 기존 ChannelListItem 대신 검색 전용 SearchResultListItem 사용

d0-neii commented on src/pages/Search.jsx:
```diff
@@ -1,9 +1,61 @@
+      return;
+    }
+
+    const result = radioChannelList.filter(({ name }) =>
```
`result`보다는 `filteredChannels`처럼 명확한 네이밍이 가독성 측면에서 더 좋을 것 같아요! 
@faber-h 희주님 생각은 어떠신지 궁금합니다😊

faber-h replied:
다은님 말씀처럼 의미가 명확하게 드러나는 네이밍이 가독성 측면에서 더 좋다는 점에 공감해요~
변경할 변수명은 `filteredChannels`도 좋고 `searchResults`도 좋을 것 같아요!

devjaeyoon replied:
말씀 주신대로 `result`보다는 의미가 더 명확하게 드러나는 `searchResults`로 추가로 `filteredList`를 `filteredChannelList`로 변수명을 변경했습니다. 검색 결과를 담는 용도임이 바로 드러나서 가독성이 좋아진 것 같아요! 
좋은 피드백 감사합니다 😊

d0-neii commented on src/pages/Search.jsx:
```diff
@@ -1,9 +1,61 @@
+    );
+
+    setFilteredList(result);
+  }, [keyword, radioChannelList]);
```
`radioChannelList`도 의존성에 포함하신 이유가 궁금합니다~
`radioChannelList`가 변하지 않는 값이라면 의존성 배열에서 제외해도 될 것 같아서요‼️

devjaeyoon replied:
좋은 질문 감사합니다!

`radioChannelList`는 현재 상태 상 자주 변하지 않는 값이긴 하지만, `zustand` 스토어에서 상태를 구독하고 있는 만큼 **언제든지 변경될 수 있는 가능성**이 있다고 생각했습니다. 또한 `eslint-plugin-react-hooks` 규칙에 따라 `radioChannelList`를 의존성 배열에 포함하지 않으면 아래와 같은 경고가 발생합니다.

> React Hook useEffect has a missing dependency: 'radioChannelList'. Either include it or remove the dependency array.

이 경고를 무시하고 제외하게 되면, 추후 `radioChannelList`에 변경이 생겼을 때 **UI가 동기화되지 않는 문제**가 발생할 수 있어 일단 포함해 두었습니다. 혹시 `radioChannelList`가 고정된 값이라는 보장이 있다면 `useMemo`로 메모이제이션하거나, 그 시점에서 의존성 제거를 고민해도 좋을 것 같습니다!

faber-h commented on src/pages/Search.jsx:
```diff
@@ -1,9 +1,61 @@
+    }
+
+    const result = radioChannelList.filter(({ name }) =>
+      name.includes(upperKeyword)
```
```suggestion
      name.toUpperCase().includes(upperKeyword)
```
대소문자 구분을 위해 `upperKeyword`를 사용해주셨네요!
그렇다면 비교를 위해 `name`도 `toUpperCase()`를 사용해야 할 것 같아요~

devjaeyoon replied:
해당 코드에 `toUpperCase()`를 사용하지 않은 이유는 저희 서비스에 라디오 채널명은 소문자가 존재하지 않아 불필요하다고 판단하여 넣지 않았는데요. 이 부분에 대해 희주님은 어떻게 생각하시나요?

devjaeyoon replied:
좋은 리뷰 감사드립니다! MBC, KBS, SBS 라디오 채널명만 생각하여 라디오 채널명은 모두 대문자일 줄 알았는데 iFM 경인방송과 같은 소문자 채널도 있었네요. 해당 내용 반영하였습니다!

faber-h commented on src/pages/Search.jsx:
```diff
@@ -1,9 +1,61 @@
+    setFilteredList(result);
+  }, [keyword, radioChannelList]);
+
+  const isEmptyInput = keyword.trim() === "";
```
`keyword`에 불필요한 공백이 있을 경우를 `trim()`으로 잘 잡아주셨네요!
다만, `trim()` 메서드가 여러 곳에서 중복 사용되고 있어서 상위에 변수로 추출해두면 코드의 일관성과 가독성을 높일 수 있을 것 같아요!
```
const trimmedKeyword = keyword.trim().toUpperCase();
```

devjaeyoon replied:
좋은 리뷰 감사드립니다. 아래 리뷰와 함께 반영하겠습니다!

faber-h commented on src/pages/Search.jsx:
```diff
@@ -1,9 +1,61 @@
+  }, [keyword, radioChannelList]);
+
+  const isEmptyInput = keyword.trim() === "";
+  const isEmptyResult = keyword && filteredList.length === 0;
```
`isEmptyInput`에서는 `trim()`을 사용하고 있는데
`isEmptyResult`에서는 원본 `keyword`를 사용하고 있어 불필요한 공백 입력 시 조건이 정확히 동작하지 않을 수 있어요.
두 조건 모두 `trim()`된 값을 기준으로 비교하면 더 명확한 동작을 보장할 수 있을 것 같아요 😊

devjaeyoon committed:
fix: 라디오 채널 검색 시 대소문자 구분 없이 검색되도록 수정

기존에는 채널명이 모두 대문자일 것으로 가정하고 keyword만 toUpperCase 처리했으나,
일부 소문자 채널명이 존재해 검색 누락 이슈가 발생함.
검색 대상인 name에도 toUpperCase()를 적용하여 대소문자 관계없이 검색 가능하도록 수정.

devjaeyoon committed:
refactor: 검색어 trim 처리 중복 제거 및 조건문 일관성 개선

- keyword.trim()을 여러 곳에서 중복 사용하던 부분을 trimmedKeyword 변수로 추출해 가독성 개선
- isEmptyInput과 isEmptyResult 모두 동일한 기준(trimmedKeyword)으로 비교하도록 수정
- 검색 조건 처리의 명확성과 유지보수성을 높이기 위한 리팩토링

devjaeyoon committed:
refactor: 검색 결과 변수명 명확하게 수정

result → searchResults, filteredList → filteredChannelList로
의미가 더 드러나도록 변수명 변경
검색 결과 관련 로직의 가독성과 유지보수성 개선

Eun0713 reviewed (approved):
LGTM💫

faber-h reviewed (approved):
LGTM 🔥
